File: sdks/python/aep/_validator.py
from __future__ import annotations

import json
import pathlib
import threading
from typing import Any

from jsonschema import Draft202012Validator
from jsonschema.exceptions import SchemaError

from ._types import CORE_EVENT_TYPES

# Schemas are bundled inside the package under aep/schemas/ so this works both
# in the monorepo (editable install) and as a standalone pip-installed package.
_SCHEMAS_DIR = pathlib.Path(__file__).parent / "schemas"
_ENVELOPE_SCHEMA_PATH = _SCHEMAS_DIR / "aep-envelope.schema.json"
_PAYLOAD_SCHEMAS_DIR = _SCHEMAS_DIR / "payloads"

_envelope_validator: Draft202012Validator | None = None
_payload_schema_cache: dict[str, Draft202012Validator | None] = {}
_validator_lock = threading.Lock()


def _get_envelope_validator() -> Draft202012Validator:
    global _envelope_validator
    if _envelope_validator is not None:
        return _envelope_validator
    with _validator_lock:
        if _envelope_validator is not None:  # re-check after acquiring lock
            return _envelope_validator
        with open(_ENVELOPE_SCHEMA_PATH, encoding="utf-8") as fh:
            schema = json.load(fh)
        _envelope_validator = Draft202012Validator(schema)
    return _envelope_validator


def _resolve_payload_validator(schema_ref: str) -> Draft202012Validator | None:
    # Fast path: check cache without lock (GIL-safe on CPython for dict reads).
    if schema_ref in _payload_schema_cache:
        return _payload_schema_cache[schema_ref]

    with _validator_lock:
        # Double-checked locking: another thread may have populated the cache
        # between the fast-path check above and acquiring the lock.
        if schema_ref in _payload_schema_cache:
            return _payload_schema_cache[schema_ref]

        # Resolve and write to cache while holding the lock so concurrent
        # callers for the same ref don't each do redundant I/O.
        result = _load_payload_schema(schema_ref)
        _payload_schema_cache[schema_ref] = result
        return result


def _load_payload_schema(schema_ref: str) -> Draft202012Validator | None:
    """Resolve *schema_ref* to a validator. Returns ``None`` if unresolvable."""
    if not _PAYLOAD_SCHEMAS_DIR.exists():
        return None

    basename = schema_ref.split("/")[-1]
    candidate = _PAYLOAD_SCHEMAS_DIR / basename
    if not candidate.exists():
        if not basename.endswith(".schema.json"):
            candidate = _PAYLOAD_SCHEMAS_DIR / (basename + ".schema.json")
        if not candidate.exists():
            return None

    try:
        with open(candidate, encoding="utf-8") as fh:
            schema = json.load(fh)
        return Draft202012Validator(schema)
    except (OSError, SchemaError, ValueError):
        return None


def _sanitize(value: Any, max_len: int = 100) -> str:
    s = str(value or "")[:max_len]
    return (
        s.replace("\\", "\\\\")
        .replace('"', '\\"')
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")
    )


def validate_event(event: dict[str, Any]) -> dict[str, bool | list[str]]:
    """Validate *event* against the AEP v0.2.0 envelope schema.

    Mirrors ``validateEvent()`` in src/validator.js.

    Returns:
        ``{"valid": bool, "errors": list[str]}``
        Entries prefixed with ``[warn]`` are non-blocking warnings.
    """
    validator = _get_envelope_validator()
    errors: list[str] = []

    schema_errors = sorted(validator.iter_errors(event), key=lambda e: list(e.absolute_path))
    for err in schema_errors:
        path = "/" + "/".join(str(p) for p in err.absolute_path) if err.absolute_path else "/"
        errors.append(f"{path} {err.message}")
    schema_ok = len(schema_errors) == 0

    type_ok = isinstance(event, dict) and event.get("type") in CORE_EVENT_TYPES
    if not type_ok:
        raw = event.get("type") if isinstance(event, dict) else None
        errors.append(f"type must be one of core v0.2 types; received '{_sanitize(raw)}'")

    # Optional payload schema validation
    payload_schema_ref: str | None = None
    if isinstance(event, dict):
        payload = event.get("payload")
        if isinstance(payload, dict):
            payload_schema_ref = payload.get("$schema")

    if payload_schema_ref:
        pv = _resolve_payload_validator(payload_schema_ref)
        if pv is not None:
            payload_errors = list(pv.iter_errors(event.get("payload", {})))
            for err in payload_errors:
                path = "payload" + (
                    "/" + "/".join(str(p) for p in err.absolute_path)
                    if err.absolute_path
                    else ""
                )
                errors.append(
                    f"{path} {err.message} (from $schema: {_sanitize(payload_schema_ref)})"
                )
        else:
            errors.append(
                f"[warn] payload.$schema '{_sanitize(payload_schema_ref)}' "
                "could not be resolved; payload accepted as-is"
            )

    blocking = [e for e in errors if not e.startswith("[warn]")]
    valid = schema_ok and type_ok and len(blocking) == 0
    return {"valid": valid, "errors": errors}

File: sdks/node/src/validator.ts
/**
 * Event validation against the AEP v0.2.0 envelope schema (+ optional payload
 * schema). Mirrors `sdks/python/aep/_validator.py` / the server's
 * `src/validator.js`. Schemas are bundled (imported JSON, inlined by the
 * bundler), so there is no runtime file I/O.
 */

import Ajv2020, { type ValidateFunction } from "ajv/dist/2020.js";
import addFormats from "ajv-formats";

import envelopeSchema from "./schemas/aep-envelope.schema.json" with { type: "json" };
import toolCalledSchema from "./schemas/payloads/tool-called.schema.json" with { type: "json" };
import { AEPEvent, CORE_EVENT_TYPES, ValidationResult } from "./types.js";

const ajv = new Ajv2020({ allErrors: true, strict: false });
addFormats(ajv);

const envelopeValidator: ValidateFunction = ajv.compile(envelopeSchema);

// Payload schemas, keyed by file basename, resolved from a payload's `$schema`.
const PAYLOAD_VALIDATORS: Record<string, ValidateFunction> = {
  "tool-called.schema.json": ajv.compile(toolCalledSchema),
};

function sanitize(value: unknown, maxLen = 100): string {
  return String(value ?? "")
    .slice(0, maxLen)
    .replace(/\\/g, "\\\\")
    .replace(/"/g, '\\"')
    .replace(/\n/g, "\\n")
    .replace(/\r/g, "\\r")
    .replace(/\t/g, "\\t");
}

function resolvePayloadValidator(schemaRef: string): ValidateFunction | null {
  const basename = schemaRef.split("/").pop() ?? schemaRef;
  if (PAYLOAD_VALIDATORS[basename]) return PAYLOAD_VALIDATORS[basename];
  if (!basename.endsWith(".schema.json")) {
    const withExt = `${basename}.schema.json`;
    if (PAYLOAD_VALIDATORS[withExt]) return PAYLOAD_VALIDATORS[withExt];
  }
  return null;
}

/**
 * Validate `event` against the AEP v0.2.0 envelope schema, the core event-type
 * list, and (if `payload.$schema` resolves to a bundled schema) the payload.
 * Returns `{ valid, errors }`; `errors` entries prefixed `[warn]` are
 * non-blocking warnings.
 */
export function validateEvent(event: AEPEvent): ValidationResult {
  const errors: string[] = [];

  const schemaOk = envelopeValidator(event);
  if (!schemaOk && envelopeValidator.errors) {
    for (const err of envelopeValidator.errors) {
      const path = err.instancePath || "/";
      errors.push(`${path} ${err.message}`);
    }
  }

  const rawType = event && typeof event === "object" ? event.type : undefined;
  const typeOk =
    !!event && typeof event === "object" && CORE_EVENT_TYPES.includes(rawType as string);
  if (!typeOk) {
    errors.push(`type must be one of core v0.2 types; received '${sanitize(rawType)}'`);
  }

  // Optional payload-schema validation, keyed off payload.$schema.
  let payloadSchemaRef: string | undefined;
  const payload = event && typeof event === "object" ? event.payload : undefined;
  if (payload && typeof payload === "object") {
    const ref = (payload as Record<string, unknown>)["$schema"];
    if (typeof ref === "string") payloadSchemaRef = ref;
  }

  if (payloadSchemaRef) {
    const pv = resolvePayloadValidator(payloadSchemaRef);
    if (pv) {
      const ok = pv(payload);
      if (!ok && pv.errors) {
        for (const err of pv.errors) {
          const path = `payload${err.instancePath || ""}`;
          errors.push(`${path} ${err.message} (from $schema: ${sanitize(payloadSchemaRef)})`);
        }
      }
    } else {
      errors.push(
        `[warn] payload.$schema '${sanitize(payloadSchemaRef)}' could not be resolved; ` +
          "payload accepted as-is",
      );
    }
  }

  const blocking = errors.filter((e) => !e.startsWith("[warn]"));
  return { valid: Boolean(schemaOk) && typeOk && blocking.length === 0, errors };
}
